### PR TITLE
[MXNET-545] Fix broken cython build

### PR DESCRIPTION
**Pull request by asitstands**
## Description ##

Fix broken cython module for ndarray and also ensure the compatibility with the recent version (0.28) of cython (#10295, #10738). See this [comment](10951#issuecomment-397852430) for the details of the changes.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by szha** [comment]
How is this tested?

**Comment by asitstands** [comment]
@szha I just checked that .cpp files are generated at `python/mxnet/cython` and .so files at `python/build/lib.linux-x86_64-3.6/mxnet/_cy3`. All unit tests are passed in my environment. I'm not sure CI has cython build. Are there any other checks do you recommend?

**Comment by asitstands** [comment]
Now, it passed all tests under `tests/python` in my environment (python2 & python3, mkl, cuda, no mkldnn, no cudnn). There were three different problems in the cython module.

1. `setup.py` is not compatible with cython 0.28. It is fixed.
1. The shared library files generated by cython have symbols defined in libmxnet.so, but the dynamic linker failed to find `libmxnet.so` unless `LD_LIBRARY_PATH` is set to the installation directory. This caused import error at runtime but the default behavior of mxnet is to try cython first and fallback to ctypes silently if an import error is encountered. So users cannot catch the error unless they set `MXNET_ENFORCE_CYTHON` env variable. This PR provides default search paths for `libmxnet.so` via rpath so that the linker can find it without `LD_LIBRARY_PATH`.
1. It looks like that the cython module for ndarray has been broken for months. It caused segfault due to the lack of a proper support of sparse arrays. This is also fixed.

**Comment by asitstands** [comment]
@szha @marcoabreu I think that it is necessary to add a cython build to CI. I'm not familiar with the CI system. Would you help?

**Comment by asitstands** [comment]
@piiswrong @szha @marcoabreu Would you review this PR?

**Comment by marcoabreu** [comment]
Sure. How can I help you?

**Comment by asitstands** [comment]
@marcoabreu Thank you. First, I'm just not sure what is the best way to add a cython build to CI. There are already many environments. Would you recommend some existing environments to add cython build? Or is it a good idea to add new environments for cython build? Second, I'm not familiar with Jenkins and docker. It looks like that I need to edit `Jenkinsfile` and files under `docker` dir. Are there any other files that I need to check?

**Comment by marcoabreu** [comment]
Depends, I'm not familiar with with cython :) Does it require its own pipeline or is it just a compile-flag we can turn on?

Everything of interest to you would be in Jenkinsfile and ci/docker, right.

**Comment by asitstands** [comment]
```bash
make

cd python
make cython

# pip install or 
export PYTHONPATH=my_installation_dir
python setup.py install --with-cython --prefix=my_installation_dir

cd ../tests/python
# run tests ...
```
I think that this is the typical way to build and run the tests. cmake build would be similar. Setting cxx/nvcc compiler flags is not required.

**Comment by marcoabreu** [comment]
Hmmm so far we don't use setup.py but just set the Python path. But I'd
welcome to see that. Feel free to change that for some tests.

What artifacts are being generated by running make in cython? We would have
to take care of copying them over as well


Deokjae Lee <notifications@github.com> schrieb am Mi., 13. Juni 2018, 22:28:

**Comment by asitstands** [comment]
I added the cython build to Python3 CentOS 7 CPU and GPU environments. I have made an intentional bug in the cython module and confirmed that it actually causes [the failures of the tests](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-10951/16/pipeline/739). I think that now this PR is completed. Please review the code. Here is the summary of the changes.

1. A bug in the `setup.py` which causes problems with the recent cython 0.28 (#10295, #10738) is fixed.
1. The cython module for the ndarray has caused segmentation faults since the introduction of the sparse array. This is fixed.
1. Now the compiled cython module has `rpath` to find `libmxnet.so`, so the deployment is easier.
1. A new makefile variable `PYTHON` to set the python executable is added (`make/config.mk`). The cython build target needs to run python but some environments have nonstandard executable name such as `python3.6`.
1. Documentation for the environment variables `MXNET_ENABLE_CYTHON` and `MXNET_ENFORCE_CYTHON` is added.
1. Documentation for the optional cython build is added to the install guide.
1. Python3 CentOS 7 CPU and Python3 CentOS 7 GPU environments in the CI now use the cython modules.

`make cython` generates two `.so` files in `mxnet/_cy3` (or `_cy2` with python 2). The exact names of the files depend on the environment. They don't need a special care for installation. It is enough just to be there.

**Comment by marcoabreu** [comment]
Sorry, forgot to press submit for my review :/

**Comment by marcoabreu** [comment]
Sorry if I missed it, but could you elaborate how we're now actually testing that cython is properly deployed and running?

@szha @eric-haibin-lin are you familiar with cython?

File: Jenkinsfile [hunk 1]
```diff
@@ -170,7 +172,7 @@ try {
           timeout(time: max_time, unit: 'MINUTES') {
             init_git()
             docker_run('centos7_cpu', 'build_centos7_cpu', false)
-            pack_lib('centos7_cpu')
+            pack_lib('centos7_cpu', mx_cython_lib)
```
[the remark this answers is not in the document]
Comment (asitstands) [reply Jenkinsfile]: Sure, I'll try soon.

File: Makefile [hunk 2]
```diff
@@ -535,7 +535,7 @@ doxygen:
 
 # Cython build
 cython:
-	cd python; python setup.py build_ext --inplace --with-cython
+	cd python; $(PYTHON) setup.py build_ext --inplace --with-cython
```
[the remark this answers is not in the document]
Comment (asitstands) [reply Makefile]: Yes, the default is just `python`. It is defined in `make/config.mk`.

**Comment by asitstands** [comment]
A way to check whether or not cython is actually used at runtime is seeing `mx.nd._internal.NDArrayBase`. It must be `mxnet._cy3.ndarray.NDArrayBase` if cython is used and `mxnet._ctypes.ndarray.NDArrayBase` otherwise.

So, if it is desirable, we could run a test like this before the unit tests run.
```bash
if [ "$(echo -e 'import mxnet as mx\nprint(mx.nd._internal.NDArrayBase)' | python)" != "<class 'mxnet._cy3.ndarray.NDArrayBase'>" ]; then
    echo "Cython is not used."
    exit 1
fi
```

**Comment by asitstands** [comment]
I removed the use of the cython modules from the Cent OS environments as @marcoabreu requested. Now "CPU: Openblas" and "GPU: Cuda 9.1" enviroments build the cython modules, and "Python 2: GPU" and "Python 3: CPU" use the cython modules to run the tests. They also run `check_cython` in the above comment to see that the cython modules are actually used.

File: ci/docker/install/ubuntu_python.sh [hunk 3]
```diff
@@ -29,5 +29,5 @@ wget -nv https://bootstrap.pypa.io/get-pip.py
 python3 get-pip.py
 python2 get-pip.py
 
-pip2 install nose cpplint==1.3.0 pylint==1.8.3 'numpy<1.15.0,>=1.8.2' nose-timer 'requests<2.19.0,>=2.18.4' h5py==2.8.0rc1 scipy==1.0.1
-pip3 install nose cpplint==1.3.0 pylint==1.8.3 'numpy<1.15.0,>=1.8.2' nose-timer 'requests<2.19.0,>=2.18.4' h5py==2.8.0rc1 scipy==1.0.1
+pip2 install nose cpplint==1.3.0 pylint==1.8.3 'numpy<1.15.0,>=1.8.2' nose-timer 'requests<2.19.0,>=2.18.4' h5py==2.8.0rc1 scipy==1.0.1 Cython
+pip3 install nose cpplint==1.3.0 pylint==1.8.3 'numpy<1.15.0,>=1.8.2' nose-timer 'requests<2.19.0,>=2.18.4' h5py==2.8.0rc1 scipy==1.0.1 Cython
```
Comment (marcoabreu) [inline ci/docker/install/ubuntu_python.sh]: Could we pin Cython, please?

Comment (asitstands) [reply ci/docker/install/ubuntu_python.sh]: I set the cython version as 0.28.2, the latest one released in April.

Comment (marcoabreu) [reply ci/docker/install/ubuntu_python.sh]: Excellent, thanks a lot!

**Comment by asitstands** [comment]
Would you please merge this PR if there is no problem?

**Comment by lupesko** [comment]
@asitstands thank you for this awesome contribution! I'm sorry to see that it has been sitting like that for such a long time...
Did you benchmark to measure perf benefits of Cython?

Tagging @anirudh2290 for review.

**Comment by lupesko** [comment]
Thanks @asitstands - this is awesome. Will try to help/facilitate getting this reviewed.

**Comment by apeforest** [comment]
@asitstands Since you initially made a mistake to set MXNET_ENABLE_CYTHON instead of MXNET_ENFORCE_CYTHON, I also find these env to be error prone. Personally, I prefer less env-based switches in the code. Do you think it makes sense to consolidate these two boolean env variables into a single int env variable such as:
```
MXNET_USE_CYTHON   0 | 1 | 2
```
0: do not use cython
1: use cython, but can fallback to ctypes
2: enforce cython

File: python/mxnet/cython/ndarray.pyx [hunk 4]
```diff
@@ -52,7 +52,7 @@ cdef class NDArrayBase:
         def __get__(self):
             return bool(self.cwritable)
 
-    def __init__(self, handle, writable=True):
+    def __init__(self, handle, writable=True, stype=-1):
```
Comment (vandanavk) [inline python/mxnet/cython/ndarray.pyx]: Is stype required here?

Comment (asitstands) [reply python/mxnet/cython/ndarray.pyx]: No, it is not required. I updated the code.

**Comment by asitstands** [comment]
@apeforest My preference is the current way of using two env variables. I think it's not so confusing. My confusion was because there was no clear documentation on the variables. So this PR adds the documentation.

File: ci/docker/runtime_functions.sh [hunk 5]
```diff
@@ -33,6 +33,18 @@ clean_repo() {
     git submodule update --init --recursive
 }
 
+check_cython() {
+    set -ex
+    local python_ver=$1
+    if [ "$(echo -e 'import mxnet as mx\nprint(mx.nd._internal.NDArrayBase.__module__)' | python${python_ver})" != "mxnet._cy${python_ver}.ndarray" ]; then
```
Comment (larroy) [inline ci/docker/runtime_functions.sh]: can we use python -c and move it before the conditional?

File: ci/docker/runtime_functions.sh [hunk 6]
```diff
@@ -343,6 +343,11 @@ build_ubuntu_cpu_openblas() {
         USE_BLAS=openblas             \
         USE_DIST_KVSTORE=1            \
         -j$(nproc)
+
+    export CC="gcc"
+    export CXX="g++"
+    make cython PYTHON=python2
```
Comment (larroy) [inline ci/docker/runtime_functions.sh]: ccache doesn't work?

**Comment by marcoabreu** [comment]
Is it really necessary to test cython on python2 and python3? Also, can't we just enable cython in an existing test-branch instead of creating a new build (while you're already changing the openblas build anyway) and two new test runs?

File: ci/docker/runtime_functions.sh (hunk 6 above)
Comment (asitstands) [reply ci/docker/runtime_functions.sh]: ccache does not work. (http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-10951/11/pipeline).

File: ci/docker/runtime_functions.sh (hunk 5 above)
Comment (asitstands) [reply ci/docker/runtime_functions.sh]: I couldn't figure out how to run multiline code with `python -c`. Anyway this kind of piping is a standard unix way. Any reason to prefer `python -c`?

**Comment by asitstands** [comment]
- We need to build cython modules for both of python2 and python3. Once the two builds are successful, I think we don't strictly need to run the tests for both. However, I think maintain the two tests in combinations with make/cmake would be a good idea.
- Currently "Python2 CPU" and "Python3 GPU" don't use cmake. One solution would be modifying "Python3 GPU" to run the test based on ''GPU: CMake'' build. So "CPU: Openblas" and "GPU: CMake" build the cython modules, and we have two unit test environments with cython: "Python2 CPU" = (make, cpu, python2) and "Python3 GPU"= (cmake, gpu, python3). I'll try this modification.
- I set MXNET_ENABLE_CYTHON=0 explicitly for all other python test environments.

**Comment by marcoabreu** [comment]
That's a great plan! Please go ahead :)

File: docs/install/index.md [hunk 7]
```diff
@@ -459,7 +459,16 @@ $ make -j $(nproc) USE_OPENCV=1 USE_BLAS=openblas
...
 
-**Step 2** Install the MXNet Python binding.
+**Step 2** Build cython modules (optional).
+
+```bash
+$ sudo apt-get install -y cython
+$ make cython # You can set the python executable with `PYTHON` flag, e.g., make cython PYTHON=python3
+```
+
+Note that you can control the use of the cython modules at runtime via the environment variables `MXNET_ENABLE_CYTHON` and `MXNET_ENFORCE_CYTHON`. See [here](/faq/env_var.html) for details.
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply docs/install/index.md]: Why not `pip install cython` ?

File: docs/install/index.md [hunk 8]
```diff
@@ -459,7 +459,16 @@ $ make -j $(nproc) USE_OPENCV=1 USE_BLAS=openblas
 $ sudo apt-get install -y python-dev python-setuptools python-pip libgfortran3
 ```
 
-**Step 2** Install the MXNet Python binding.
+**Step 2** Build cython modules (optional).
```
Comment (sandeep-krishnamurthy) [inline docs/install/index.md]: Can we please add instructions for Mac and Windows users?

File: ci/docker/runtime_functions.sh (hunk 5 above)
Comment (larroy) [reply ci/docker/runtime_functions.sh]: I guess doesn't make a difference. Can we store in a variable with propper name? It's difficult to read inside the if condition.

File: docs/install/index.md (hunk 8 above)
Comment (asitstands) [reply docs/install/index.md]: Unfortunately I'm not familiar with Windows/Mac and also don't have environments to test. I think this needs help from others.

Comment (lupesko) [reply docs/install/index.md]: @sandeep-krishnamurthy maybe we can start with Linux only, and leverage other's help to support Mac and Windows?

Comment (sandeep-krishnamurthy) [reply docs/install/index.md]: Yes we can iteratively update.

**Comment by asitstands** [comment]
The builds of the cython modules in "CPU: Openblas" and "GPU: CMake" are successful. Tests in "Python2 CPU" = (make, cpu, python2) are also ok. In "Python3 GPU"= (cmake, gpu, python3), however, all tests are passed and the final report is printed, but a segmentation fault occurs at the end [(log)](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-10951/52/pipeline). I tried the same tests with ctypes, instead of the cython modules, but it makes no difference [(log)](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-10951/48/pipeline). So it may not be a problem of the cython modules. However, everything is ok with the same tests based on a make build (instead of cmake) [(log)](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-10951/50/pipeline/878). So there may be some difference between the make and cmake builds, but I couldn't figure out what is the actual difference that makes the segfault. The two builds are 

```
cmake \
    -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
    -DCMAKE_C_COMPILER_LAUNCHER=ccache \
    -DUSE_CUDA=1               \
    -DUSE_CUDNN=1              \
    -DUSE_MKLML_MKL=0          \
    -DUSE_MKLDNN=0             \
    -DUSE_DIST_KVSTORE=1       \
    -DCMAKE_BUILD_TYPE=Release \
    -G Ninja                   \
    /work/mxnet
```
and
```bash
make \
    USE_CUDA=1                    \
    USE_CUDNN=1                   \
    USE_CUDA_PATH=/usr/local/cuda \
    USE_MKLML_MKL=0               \
    USE_MKLDNN=0                  \
    USE_DIST_KVSTORE=1            \
    -j$(nproc)
```
I have disabled ccache and cudnn, but they didn't make any difference. I also tried to reproduce the segfault in my local environment but failed. Any advice would be helpful.

File: ci/Jenkinsfile_utils.groovy [hunk 9]
```diff
@@ -57,20 +57,20 @@ def init_git_win() {
 }
 
 // pack libraries for later use
-def pack_lib(name, libs) {
+def pack_lib(name, libs=mx_lib) {
```
Comment (marcoabreu) [inline ci/Jenkinsfile_utils.groovy]: Unrelated changes?

Comment (asitstands) [reply ci/Jenkinsfile_utils.groovy]: Oops, it was a careless copy and paste. Now I removed the default value.

**Comment by vandanavk** [comment]
@mxnet-label-bot [pr-awaiting-testing] To resolve branch conflicts and CI failures.

This PR is waiting on https://github.com/apache/incubator-mxnet/pull/12160

@marcoabreu for review

**Comment by vrakesh** [comment]
@marcoabreu Requesting to review this PR.

**Comment by asitstands** [comment]
@vrakesh I'm still waiting on #12160. I'll update this PR to resolve the conflict and make sure that it passes the tests after #12160 is addressed.

**Comment by Roshrini** [comment]
@lebeg This PR is waiting on #12160  It would be great if you can follow up on that.

**Comment by ankkhedia** [comment]
@lebeg Request you to please follow up on #12160 so that this PR can move forward

**Comment by asitstands** [comment]
#12160 looks like delayed too much. The issue is that the cmake build with llvm openmp causes a segfault in the CI environment. The segfault also occurs without cython. So the issue is not directly related to this PR. Of course, it would be better if we could test the cython build with cmake. However, in current situation, I think that we need to separate this PR with the cmake build issue. I removed the testing of the cython modules with cmake. 

'CPU: Openblas' and 'GPU: CUDA9.1+cuDNN7' build cython module with make. 'Python2: CPU' and 'Python3: GPU' runs unit tests with the cython modules. 'GPU: CMake' still builds cython modules with cmake but there is no test with them.

@marcoabreu @larroy @lebeg

**Comment by kalyc** [comment]
@mxnet-label-bot update [pr-awaiting-review]

**Comment by asitstands** [comment]
Any discussion on the last change? Or can we get this PR merged?

**Comment by marcoabreu** [comment]
I'm very sorry that this PR takes so long to get merged. I will ask some colleagues if they are able to help you reviewing this.

Also, thanks a lot for the changes to the new Jenkinsfile layout!

**Comment by vandanavk** [comment]
@marcoabreu is this PR good to go?

**Comment by larroy** [comment]
Do we have a bug tracking the segfault with openmp? 

The PR looks ok to me, even though I'm not familiar with cython. I think if you add some motivation to the description of the PR explaining why is this needed and some measure if it makes things faster would help the reviewers understand better the PR.

Another question is, this option is not available in the CMake system, shall we add it, but mark it as broken so we don't have divergent builds between Make & CMake?

**Comment by asitstands** [comment]
@larroy This PR just fixes a broken existing feature. Why do we need cython? We may have some performance gain. Here is the copy of a quick benchmark result I reported before. For a quick benchmark I have run `example/gluon/image_classification.py` with `--model=resnet50_v1 --dataset=cifar10` on a machine with two Xeon E5-2680 and 1080ti. The mean time consumption per epoch was 72.2s with cython and 78.6s without cython. So the the running time with cython is about 8% shorter. With `--model=resnet101_v1`, the gain was about 10% (135.3s vs 150.2s).

For cmake, I added `BUILD_CYTHON_MODULES` option. 'GPU: CMake' environment builds the cython modules via cmake but no unit test runs with the artifact from the build. I left a comment about the llvm openmp issue at the unit test 'Python3: GPU'. 

https://github.com/apache/incubator-mxnet/blob/adce15e25afdf3a47f5dc26bda9db7371aa4882b/ci/docker/runtime_functions.sh#L665-L682

https://github.com/apache/incubator-mxnet/blob/10d817cea15ec0b3253f3b78a30052f764931026/ci/jenkins/Jenkins_steps.groovy#L599-L612

**Comment by sandeep-krishnamurthy** [comment]
@larroy / @marcoabreu - Is this good to go?

**Comment by Roshrini** [comment]
@marcoabreu @larroy Could you please take a look/merge this PR?

**Comment by stu1130** [comment]
@asitstands could you resolve the conflict? (maybe rebase it) Thanks

**Comment by larroy** [comment]
I'm not an expert in cython but the changes look sane. I think we can merge this at least tentatively and observe. Since the author did some benchmarking and there are gains we need to keep an eye on performance numbers. I would be in favor of merging and if there's unexpected ramifications we can revert.

File: ci/Jenkinsfile_utils.groovy [hunk 10]
```diff
@@ -67,7 +67,7 @@ def pack_lib(name, libs, include_gcov_data = false) {
   sh returnStatus: true, script: """
 set +e
 echo "Packing ${libs} into ${name}"
-echo ${libs} | sed -e 's/,/ /g' | xargs md5sum
+for i in \$(echo ${libs} | sed -e 's/,/ /g'); do md5sum \$i; done
```
Comment (larroy) [inline ci/Jenkinsfile_utils.groovy]: Isn't this the same?

Comment (asitstands) [reply ci/Jenkinsfile_utils.groovy]: This is to handle glob patterns in `$libs`. The names of the shared library files produced by cython depend on the python version and platform. Using glob pattterns is convenient.

**Comment by ankkhedia** [comment]
@marcoabreu Could you please review/merge this PR?

**Comment by szha** [comment]
@marcoabreu given that it has been open for a while, @marcoabreu could you take a look again and see if the CI-related changes are still up to date? Thanks.

**Comment by sl1pkn07** [comment]
anyone can rebase this pr? i can't apply it

for this https://github.com/apache/incubator-mxnet/issues/10738

**Comment by vandanavk** [comment]
@marcoabreu is this PR good to go?

**Comment by sl1pkn07** [comment]
i cant apply this PR with the patch method in git HEAD

https://patch-diff.githubusercontent.com/raw/apache/incubator-mxnet/pull/10951.patch

anyone can help me?

fails patch about JenkinsFile and other errors

greetings

**Comment by sl1pkn07** [comment]
ok `git pull --no-edit origin pull/10951/head` do the trick

**Comment by asitstands** [comment]
Can we get this PR merged? I'm not sure what prevents the merging. Anyway MXNet does not use cython modules unless the env var MXNET_USE_CYTHON is set in runtime. So this PR does not change any behavior of MXNet unless the env var is set. Merging this PR is relatively safe and we could observe its effect after merging. Or we could completely remove the cython modules. They exist but are broken. Keep the broken things would not be a good idea.

**Comment by sl1pkn07** [comment]
broken again if use with master

**Comment by larroy** [comment]
Agree we should move this PR forward, or do something about it. Has been open for too long.

**Comment by larroy** [comment]
Do we need to rebase it?

**Comment by larroy** [comment]
@asitstands could you rebase so we try to get this merged?

**Comment by asitstands** [comment]
I'm not sure why this PR shows unrelated commits with hundreds of changed files. My own repository shows only 13 files actually changed by this PR. I found that the target branch of this PR has been changed to `apache:cython_fix` from `apache:master`. Any relation with this? @szha

**Comment by szha** [comment]
@asitstands as this PR has been around for quite a while I changed it to a feature branch with the hope to merge it there so that others may carry on this work. Since you're working on this now, it's no longer necessary so I changed it back. It should be ready for review after another rebase.

**Comment by asitstands** [comment]
Now this PR has no confliction with the main branch and passes the tests. Please review whether it is OK to be merged.
